### Q.1.1 Computing

In [2]:
#1 How many seconds are there in year?
seconds_in_year = 365 * 24 * 60 * 60
print(seconds_in_year)

31536000


- There are 31536000 seconds in one year.

In [3]:
#2 How many seconds in a typical human lifetime? Use the previous computed seconds in year.
seconds_lifetime = 80 * seconds_in_year
print(seconds_lifetime)

2522880000


- Typical human lifetime is about 80 years. So, most people live 2522880000 seconds.

In [4]:
#3 What is your age in seconds? Compute this based on your age.
my_age_in_seconds = 21 * seconds_in_year
print(my_age_in_seconds)

662256000


- I am 21 years old. In other words, I am 662256000 seconds old.

In [5]:
#4 Create a logical variable that is true if you are more than 700M seconds old. Use logical operator, not if/else!
print(bool(my_age_in_seconds > 700000000))

False


In [6]:
#5 Compute the following mathematical formulas:
p = 0.4
p * (1 - p)

0.24

In [7]:
a = 3
x0 = 2
x = 2.5
a / x0 * (1 + x / x0)**(-a - 1)

0.058527663465935076

### Q.1.2 Lists

In [8]:
#1 Create a list 'movies' that contains the names of at least six movies you like.
movies = ['Spiderman', 'Kissing Booth', 'Inception', 'Up', 'La La Land', 'Titanic']
print(movies)

['Spiderman', 'Kissing Booth', 'Inception', 'Up', 'La La Land', 'Titanic']


In [9]:
#2 Using indexing and slicing, Create a list of three first movies in the list.
movies[:3]

['Spiderman', 'Kissing Booth', 'Inception']

In [10]:
#3 Use slicing and list concentration to create a list of the first two and the last two movies.
movies[0:2] + movies[4:]

['Spiderman', 'Kissing Booth', 'La La Land', 'Titanic']

In [11]:
#4 Use slicing to list the movies in the opposite order.
movies[::-1]

['Titanic', 'La La Land', 'Up', 'Inception', 'Kissing Booth', 'Spiderman']

### Q.1.3 Loops

In [12]:
#1 Create a list 'numbers' that is the numbers 70 through 79 in the following manner.
numbers = []
for i in range(70,80):
    numbers.append(i)
print(numbers)

[70, 71, 72, 73, 74, 75, 76, 77, 78, 79]


In [13]:
#2 Use a loop to compute the mean value of the list.
sum = 0
for i in numbers:
    sum += i
print(sum / 10)

74.5


In [14]:
#3 Use loop to compute sum of all integers from 1 till 100.
sum = 0
for i in range(1,101):
    sum += i
print(sum)

5050


In [15]:
#4 Assign people to seat.
people = ["Adam", "Ashin", "Inukai", "Tanaka", "Ikki"]
seat = ["33", "12", "45", "2", "17"]
assigned = []
for i in range(5):
    assigned.append(people[i] + ": " + seat[i])
print(assigned)

['Adam: 33', 'Ashin: 12', 'Inukai: 45', 'Tanaka: 2', 'Ikki: 17']


### Q.1.4 Dicts

In [16]:
#1 Count the number of each word in a sentence.
def countWords(individual):
    word_count = {}
    for word in individual:
        if not word in word_count:
            word_count[word] = individual.count(word)
    return word_count

sentence = """The final structure we discuss here is set.
It models the set in the mathematical sense, i.e. it is an unordered collection that contains only one copy of each element. 
Unordered means that looping over the elements extracts those in an unpredictable order, 
and it does not support positional access either."""
individual = sentence.lower().split()

word_count = countWords(individual)
print(word_count)

{'the': 4, 'final': 1, 'structure': 1, 'we': 1, 'discuss': 1, 'here': 1, 'is': 2, 'set.': 1, 'it': 3, 'models': 1, 'set': 1, 'in': 2, 'mathematical': 1, 'sense,': 1, 'i.e.': 1, 'an': 2, 'unordered': 2, 'collection': 1, 'that': 2, 'contains': 1, 'only': 1, 'one': 1, 'copy': 1, 'of': 1, 'each': 1, 'element.': 1, 'means': 1, 'looping': 1, 'over': 1, 'elements': 1, 'extracts': 1, 'those': 1, 'unpredictable': 1, 'order,': 1, 'and': 1, 'does': 1, 'not': 1, 'support': 1, 'positional': 1, 'access': 1, 'either.': 1}


### Q.1.5 Functions and data transformations

In [17]:
#1 Change Hours-Minutes to Hours-Fractions of hours.
def calculator(num):
    minute = num % 100
    hour = num // 100

    num_in_minute = hour * 60 + minute
    return (num_in_minute / 60)

calculator(1015), calculator(345)

(10.25, 3.75)

### Q.2.1 Setup

In [18]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [19]:
#1 Load the data.
nyc_flights = pd.read_csv("../data/nycflights13.csv.bz2", sep="\t")

In [20]:
#2 Basic checks
nyc_flights.shape

(336776, 19)

- There are 19 columns and 336776 rows in the data. 

In [21]:
nyc_flights.columns

Index(['year', 'month', 'day', 'dep_time', 'sched_dep_time', 'dep_delay',
       'arr_time', 'sched_arr_time', 'arr_delay', 'carrier', 'flight',
       'tailnum', 'origin', 'dest', 'air_time', 'distance', 'hour', 'minute',
       'time_hour'],
      dtype='object')

In [22]:
nyc_flights.head()

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0             515        2.0     830.0   
1  2013      1    1     533.0             529        4.0     850.0   
2  2013      1    1     542.0             540        2.0     923.0   
3  2013      1    1     544.0             545       -1.0    1004.0   
4  2013      1    1     554.0             600       -6.0     812.0   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1             830       20.0      UA    1714  N24211    LGA  IAH     227.0   
2             850       33.0      AA    1141  N619AA    JFK  MIA     160.0   
3            1022      -18.0      B6     725  N804JB    JFK  BQN     183.0   
4             837      -25.0      DL     461  N668DN    LGA  ATL     116.0   

   distance  hour  minute             time_hour  
0      1400     5      15  2013-01-01T10:00:00Z  
1      1416     5      29  2013-01-01T10:00:00Z  
2      1089     5      40  2013-01-01T10:00:00Z  
3      1576     5      45  2013-01-01T10:00:00Z  
4       762     6       0  2013-01-01T11:00:00Z

### Q.2.2 Basic data exploration

In [23]:
#1 How many NYC airports are included in this data? Which airports are these?
nyc_flights.origin.nunique()

3

In [24]:
nyc_flights.origin.unique()

array(['EWR', 'LGA', 'JFK'], dtype=object)

- There are total three airports in this data: EWR, LGA, and JFK.

In [25]:
#2 Into how many airports did the airlines fly from NYC in 2013?
nyc_flights[nyc_flights['year'] == 2013].dest.nunique()

105

- The airlines fly from NYC into 105 airports in 2013.

In [26]:
#3 How many flights were there from NYC to Seattle(SEA)?
nyc_flights[nyc_flights['dest'] == 'SEA'].count()['dest']

3923

- There were 3923 flights from NYC to SEA.

In [27]:
#4 Were there any flights from NYC to Spokane(GEG)?
nyc_flights[nyc_flights['dest'] == 'GEG'].count()['dest']

0

- There were no flight from NYC to GEG.

### Q.2.3 Compute new variables

In [28]:
#1 Create new column for the flight data-flight duration in hours.
nyc_flights['duration'] = nyc_flights.arr_time - nyc_flights.dep_time

In [59]:
nyc_flights.sample(5)

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
91416   2013     12    9    2055.0            1830      145.0    2253.0   
248098  2013      6   28    1148.0            1030       78.0    1507.0   
295391  2013      8   17     640.0             638        2.0     907.0   
315190  2013      9    7    1620.0            1455       85.0    1741.0   
89541   2013     12    7    1700.0            1640       20.0    2003.0   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
91416             2046      127.0      EV    5314  N615QX    EWR  DTW   
248098            1345       82.0      AA     179  N327AA    JFK  SFO   
295391             926      -19.0      UA     441  N579UA    EWR  SEA   
315190            1630       71.0      MQ    3231  N639MQ    JFK  PIT   
89541             1954        9.0      DL    1854  N332NB    LGA  FLL   

        air_time  distance  hour  minute             time_hour  duration  
91416       91.0       488    18      30  2013-12-09T23:00:00Z     198.0  
248098     333.0      2586    10      30  2013-06-28T14:00:00Z     359.0  
295391     311.0      2402     6      38  2013-08-17T10:00:00Z     267.0  
315190      62.0       340    14      55  2013-09-07T18:00:00Z     121.0  
89541      163.0      1076    16      40  2013-12-07T21:00:00Z     303.0

In [30]:
#2 What is the minimum value of the duration you computed? Explain why you get negative values!
nyc_flights.duration.min()

-2346.0

- There are some negative values of duration since some arr_time is smaller than some dep_time. 

### Q.2.4 Flight Delays

In [61]:
#1 Create a new data frame, a subset of the flights data that only contains two variables: departure delay and arrival delay.
new_flights = nyc_flights[['dep_delay', 'arr_delay']]
new_flights.head()

dep_delay  arr_delay
0        2.0       11.0
1        4.0       20.0
2        2.0       33.0
3       -1.0      -18.0
4       -6.0      -25.0

In [62]:
#2 Print a small random sample of the delay data frame.
new_flights.sample(5)

dep_delay  arr_delay
39778        79.0       95.0
171154       18.0        6.0
153206        4.0       -4.0
299669       30.0       24.0
319629        4.0        4.0

In [65]:
#3 What is the data type of the flight delay value?
new_flights.dtypes

dep_delay    float64
arr_delay    float64
dtype: object

In [75]:
#4 Filter to keep the "slow flights": cases where the plane leaves early but arrives late. How many cases do you have?
new_flights[(new_flights.dep_delay < 0) & (new_flights.arr_delay > 0)].shape

(35648, 3)

- There are 35648 cases where plane leaves early but arrives late.

In [78]:
#5 What is the longest "extra flight time"?
new_flights['extra_flight'].copy = new_flights.arr_delay - new_flights.dep_delay
new_flights.head()

dep_delay  arr_delay  extra_flight
0        2.0       11.0           9.0
1        4.0       20.0          16.0
2        2.0       33.0          31.0
3       -1.0      -18.0         -17.0
4       -6.0      -25.0         -19.0

In [79]:
new_flights.extra_flight.max()

196.0

- The longest extra flight time is 196.0.

In [83]:
#6 How many of flights were delayed (positive arrival delay)?
new_flights[new_flights.arr_delay > 0].shape

(133004, 3)

- There are 133004 delayed flights.

In [103]:
#7 What percentage of flights were delayed by more than 15 minutes at arrival? 
(new_flights.arr_delay > 15).mean()

0.2305092999501152

- About 23.05% of flights were delayed by more than 16 minutes.

### How much time did you spend?

- I spent about 6 hours.